src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Components.Endpoints;
+
+/// <summary>
+/// A component that describes a location in prerendered output where client-side code
+/// should insert an interactive component.
+/// </summary>
+internal class SSRRenderModeBoundary : IComponent
+{
+    private readonly Type _componentType;
+    private readonly IComponentRenderMode _renderMode;
+    private readonly bool _prerender;
+    private RenderHandle _renderHandle;
+    private IReadOnlyDictionary<string, object?>? _latestParameters;
+
+    public SSRRenderModeBoundary(Type componentType, IComponentRenderMode renderMode)
+    {
+        _componentType = componentType;
+        _renderMode = renderMode;
+        _prerender = renderMode switch
+        {
+            ServerRenderMode mode => mode.Prerender,
+            WebAssemblyRenderMode mode => mode.Prerender,
+            AutoRenderMode mode => mode.Prerender,
+            _ => throw new ArgumentException($"Server-side rendering does not support the render mode '{renderMode}'.", nameof(renderMode))
+        };
+    }
+
+    public void Attach(RenderHandle renderHandle)
+    {
+        _renderHandle = renderHandle;
+    }
+
+    public Task SetParametersAsync(ParameterView parameters)
+    {
+        // We have to snapshot the parameters because ParameterView is like a ref struct - it can't escape the
+        // call stack because the underlying buffer may get reused. This is enforced through a runtime check.
+        _latestParameters = parameters.ToDictionary();
+
+        if (_prerender)
+        {
+            _renderHandle.Render(Prerender);
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private void Prerender(RenderTreeBuilder builder)
+    {
+        builder.OpenComponent(0, _componentType);
+
+        foreach (var (name, value) in _latestParameters!)
+        {
+            builder.AddComponentParameter(1, name, value);
+        }
+
+        builder.CloseComponent();
+    }
+
+    public (ServerComponentMarker?, WebAssemblyComponentMarker?) ToMarkers(HttpContext httpContext)
```

Comment:
Just so I'm understanding this correctly - in the case of the "Auto" render mode, we'll emit markers of both types? And on the client side, we'll have some logic to detect if any given component has multiple markers and treat that as "auto" mode?

Reply:
Yes, in the auto case, the client has to make the choice about what mode to use (at least, if it's based on "is the WebAssembly .NET runtime ready to start"). We are free to change the format of these markers and perhaps come up with a format that merges the server and WebAssembly information into a single structure, but this PR is only about emitting markers in the existing format. We should design a new format, if we want one, based on the needs of the client-side code when we implement that.